server/commands/write.go

```diff
@@ -87,6 +90,58 @@ func (c *WriteCommand) validateTuplesets(ctx context.Context, req *openfgapb.Wri
 	return nil
 }
 
+// validateTypes makes sure that when writing a tuple, the types are compatible.
+// 1. If the tuple is of the form (person:bob, reader, doc:budget), then the type "doc", relation "reader" allows type "person".
+// 2. If the tuple is of the form (group:abc#member, reader, doc:budget), then the type "doc", relation "reader" must allow type "group", relation "member".
+// 3. If the tuple is of the form (*, reader, doc:budget), we allow it only if the type "doc" relation "reader" allows at least one type (with no relation)
+func (c *WriteCommand) validateTypes(ctx context.Context, store string, authorizationModelID string, tk *openfgapb.TupleKey, dbCallsCounter utils.DBCallCounter) error {
+	objectType, _ := tupleUtils.SplitObject(tk.GetObject()) // e.g. "doc"
+
+	dbCallsCounter.AddReadCall()
+	typeDefinition, err := c.datastore.ReadTypeDefinition(ctx, store, authorizationModelID, objectType)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return &tupleUtils.TypeNotFoundError{TypeName: objectType}
+		}
+		return err
+	}
+	relations := typeDefinition.GetMetadata().GetRelations()
```

Comment:
Let's hold off on this PR a bit until the `typesystem` package gets merged in from the other PRs. A lot of this code can be reduced to helpers that will be in that package.